### PR TITLE
fix: make cleanupOldChecks honor tier retention

### LINKED ISSUE #108: [P0] Fix data retention: cleanupOldChecks must respect tier (Vital = 90 days)
## Problem
\`convex/monitoring.ts:cleanupOldChecks\` deletes all checks older than 30 days unconditionally for every monitor, regardless of the owner's subscription tier.

Vital tier users pay for 90-day check history (advertised on pricing page, enforced in \`lib/tiers.ts\`). The cleanup cron actively deletes their paid-for data every day. This is a billing integrity bug: users are paying for a feature the system is quietly destroying.

The fix requires tier-aware retention: look up the monitor owner's tier, use 30 days for Pulse and 90 days for Vital.

## Context
- **Vision**: Set-and-forget uptime monitoring with beautiful public status pages for indie hackers
- **Related**: None
- Affects every Vital tier user. If a Vital user checks their 45-day check history and finds it empty, there is no recovery — the data is gone.

## Acceptance Criteria
- [ ] Given a Vital user's monitor with checks from 45 days ago, when the daily cleanup runs, then those checks are NOT deleted
- [ ] Given a Pulse user's monitor with checks from 45 days ago, when the daily cleanup runs, then those checks ARE deleted
- [ ] Given a monitor with no subscription (trial, cancelled), when cleanup runs, then it uses the Pulse threshold (30 days) as a safe default
- [ ] Given the cleanup cron, when it runs, then it processes all monitors without hitting Convex read limits (existing pagination is preserved)

## Affected Files
- `convex/monitoring.ts` — `cleanupOldChecks` internal action: look up monitor owner's subscription tier and use tier-appropriate cutoff
- `convex/subscriptions.ts` — may need an internal query to get retention days by userId
- `lib/tiers.ts` — verify `historyDays` is defined per tier (Pulse: 30, Vital: 90); use as source of truth

## Verification
```bash
bun test --grep "cleanup"
bun lint && bun type-check
# Manual verification: check that a Vital monitor's 60-day-old checks still exist after cleanup
```

## Boundaries
- Do NOT change the pagination/loop structure of the existing cleanup — it's correct, only the cutoff needs to change
- Do NOT retroactively restore deleted data — this is a forward-fix only

## Approach
```typescript
// In cleanupOldChecks, after fetching monitors to clean:
const subscription = await ctx.runQuery(internal.subscriptions.getByUserId, {
  userId: monitor.userId
});
const tier = subscription?.tier ?? "pulse";
const retentionDays = tier === "vital" ? 90 : 30;
const cutoff = Date.now() - retentionDays * 24 * 60 * 60 * 1000;
```

Pull the retention days value from \`lib/tiers.ts\`'s \`historyDays\` field rather than hard-coding numbers.

## Overview
```mermaid
flowchart TD
    A[Daily cleanup cron] --> B[Fetch batch of monitors]
    B --> C{For each monitor}
    C --> D[Look up owner subscription tier]
    D --> E{Tier?}
    E -->|vital| F[cutoff = now - 90 days]
    E -->|pulse / default| G[cutoff = now - 30 days]
    F --> H[Delete checks older than cutoff]
    G --> H
```

---
Created by `/groom`

### PR DESCRIPTION
**Pull request by phrazzld**
## Summary
Fixes tier-blind retention in daily check cleanup so paid Vital history is no longer deleted at 30 days.

Closes #108.

## Changes
- Added `historyDays` to shared server/client tier limits in `convex/constants.ts`.
- Updated `convex/monitoring.ts::cleanupOldChecks` to:
  - fetch candidate checks older than Pulse threshold,
  - resolve monitor owner subscription tier/status,
  - delete only checks beyond the owner-specific retention window.
- Added `getMonitorsByIds` internal query to avoid per-check monitor lookups.
- Added cleanup retention tests in `convex/__tests__/monitoring.test.ts` for:
  - active Vital keeps 45-day checks,
  - Pulse deletes 45-day checks,
  - non-active subscriptions default to Pulse retention.
- Removed duplicated `historyDays` literals from `lib/tiers.ts` by relying on shared tier constants.

## Acceptance Criteria
- [x] Given a Vital user's monitor with checks from 45 days ago, cleanup does not delete them.
- [x] Given a Pulse user's monitor with checks from 45 days ago, cleanup deletes them.
- [x] Given a non-active subscription context, cleanup uses Pulse retention as default.
- [x] Cleanup keeps existing bounded batch loop behavior and processes candidates in batches.

## Manual QA
1. `bun lint`
2. `bun type-check`
3. `bunx vitest run convex/__tests__/monitoring.test.ts`
4. `git push -u origin fix/108-tier-aware-check-retention` (runs full pre-push test suite)
5. Expected:
- lint/type-check pass
- monitoring retention tests pass
- pre-push `vitest run` passes

## What Changed
```mermaid
graph TD
  A[cleanupOldChecks] --> B[get checks older than Pulse cutoff]
  B --> C[load check monitors]
  C --> D[load owner subscription]
  D --> E{active vital?}
  E -->|yes| F[retention = 90 days]
  E -->|no| G[retention = 30 days]
  F --> H[delete only checks older than monitor cutoff]
  G --> H
```

## Before / After
Before: cleanup deleted all checks older than 30 days, including Vital users who should retain 90 days.

After: cleanup applies retention per monitor owner (active Vital => 90 days, otherwise Pulse default 30 days), preserving paid history while still pruning old data.

## Test Coverage
- `convex/__tests__/monitoring.test.ts`
  - `cleanupOldChecks > keeps 45-day checks for active Vital subscriptions`
  - `cleanupOldChecks > deletes 45-day checks for Pulse subscriptions`
  - `cleanupOldChecks > defaults to Pulse retention for non-active subscriptions`
- Full repo tests executed by pre-push hook (`vitest run`): 58 files, 987 tests passing.

## Notes / Gaps
- Current fallback behavior for non-active subscription status is intentionally conservative (Pulse retention).

## Summary by CodeRabbit

* **New Features**
  * Implemented tier-aware data retention policies: Vital subscriptions now retain monitoring data for 90 days, while Pulse subscriptions retain data for 30 days. Old checks are automatically cleaned up based on subscription tier.

* **Tests**
  * Added comprehensive test suite validating retention behavior across subscription tiers, including edge cases for active and non-active subscriptions.

* **Documentation**
  * Added delivery retro documenting implementation patterns and key learnings from this feature.

### CODE REVIEW

File: convex/monitoring.ts [hunk 1]
```diff
@@ -403,20 +420,72 @@ export const cleanupOldChecks = internalAction({
...
+          getCheckRetentionDays(subscription),
+        );
+      }
+
+      const now = Date.now();
+      const checksToDelete = oldChecks.filter((check) => {
+        const monitor = monitorsById.get(check.monitorId);
+        if (!monitor) {
+          return false;
+        }
```
[the remark this answers is not in the document]
Comment (phrazzld) [reply convex/monitoring.ts]: Classification: bug
Severity: high
Decision: fix now. Orphaned checks should not be retained indefinitely.
Change: Cleanup now treats missing monitor rows as orphaned checks and deletes them (, commit 24af278).
Verification: Added/ran  in ; 
 RUN  v4.0.18 /Users/phaedrus/Development/heartbeat

stdout | convex/__tests__/monitoring.test.ts > updateMonitorStatus > third failure sets counter to 3
[Notification] No settings for user user_test, skipping

stdout | convex/__tests__/monitoring.test.ts > updateMonitorStatus > success resets failures even from high count
[Notification] No settings for user user_test, skipping

stdout | convex/__tests__/monitoring.test.ts > updateMonitorStatus > success resets failures even from high count
[Notification] No settings for user user_test, skipping

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > keeps 45-day checks for active Vital subscriptions
[Cleanup] Starting tier-aware check cleanup...

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > keeps 45-day checks for active Vital subscriptions
[Cleanup] Batch 1: deleting 0/1 eligible checks
[Cleanup] Complete: deleted 0 checks total

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > deletes 45-day checks for Pulse subscriptions
[Cleanup] Starting tier-aware check cleanup...

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > deletes 45-day checks for Pulse subscriptions
[Cleanup] Batch 1: deleting 1/1 eligible checks

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > deletes 45-day checks for Pulse subscriptions
[Cleanup] Complete: deleted 1 checks total

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > defaults to Pulse retention for non-active subscriptions
[Cleanup] Starting tier-aware check cleanup...

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > defaults to Pulse retention for non-active subscriptions
[Cleanup] Batch 1: deleting 1/1 eligible checks

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > defaults to Pulse retention for non-active subscriptions
[Cleanup] Complete: deleted 1 checks total

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > continues past non-deletable page to delete eligible pulse checks
[Cleanup] Starting tier-aware check cleanup...

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > continues past non-deletable page to delete eligible pulse checks
[Cleanup] Batch 1: deleting 0/1000 eligible checks

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > continues past non-deletable page to delete eligible pulse checks
[Cleanup] Batch 2: deleting 1/2 eligible checks

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > continues past non-deletable page to delete eligible pulse checks
[Cleanup] Complete: deleted 1 checks total

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > deletes orphaned checks when monitor was removed
[Cleanup] Starting tier-aware check cleanup...

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > deletes orphaned checks when monitor was removed
[Cleanup] Batch 1: deleting 1/1 eligible checks

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > deletes orphaned checks when monitor was removed
[Cleanup] Complete: deleted 1 checks total

 ✓ convex/__tests__/monitoring.test.ts (21 tests) 167ms

 Test Files  1 passed (1)
      Tests  21 passed (21)
   Start at  20:46:38
   Duration  562ms (transform 110ms, setup 47ms, import 48ms, tests 167ms, environment 223ms).

File: convex/monitoring.ts [hunk 2]
```diff
@@ -403,20 +420,72 @@ export const cleanupOldChecks = internalAction({
...
       console.log(
-        `[Cleanup] Batch ${batchNumber}: deleting ${oldChecks.length} checks`,
+        `[Cleanup] Batch ${batchNumber}: deleting ${checksToDelete.length}/${oldChecks.length} eligible checks`,
       );
 
+      if (checksToDelete.length === 0) {
+        console.log(
+          `[Cleanup] Complete: no checks eligible for deletion in current window (${totalDeleted} deleted total)`,
+        );
+        return;
```
[the remark this answers is not in the document]
Comment (phrazzld) [reply convex/monitoring.ts]: Classification: bug
Severity: high
Decision: fix now. Early completion on a single non-deletable page could skip later eligible records.
Change: Implemented cursor pagination for cleanup () and continue-scanning behavior across pages, with run cap + reschedule preserved (, commit 24af278).
Verification: Added/ran ; 
 RUN  v4.0.18 /Users/phaedrus/Development/heartbeat

stdout | convex/__tests__/monitoring.test.ts > updateMonitorStatus > third failure sets counter to 3
[Notification] No settings for user user_test, skipping

stdout | convex/__tests__/monitoring.test.ts > updateMonitorStatus > success resets failures even from high count
[Notification] No settings for user user_test, skipping

stdout | convex/__tests__/monitoring.test.ts > updateMonitorStatus > success resets failures even from high count
[Notification] No settings for user user_test, skipping

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > keeps 45-day checks for active Vital subscriptions
[Cleanup] Starting tier-aware check cleanup...

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > keeps 45-day checks for active Vital subscriptions
[Cleanup] Batch 1: deleting 0/1 eligible checks
[Cleanup] Complete: deleted 0 checks total

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > deletes 45-day checks for Pulse subscriptions
[Cleanup] Starting tier-aware check cleanup...

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > deletes 45-day checks for Pulse subscriptions
[Cleanup] Batch 1: deleting 1/1 eligible checks

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > deletes 45-day checks for Pulse subscriptions
[Cleanup] Complete: deleted 1 checks total

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > defaults to Pulse retention for non-active subscriptions
[Cleanup] Starting tier-aware check cleanup...

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > defaults to Pulse retention for non-active subscriptions
[Cleanup] Batch 1: deleting 1/1 eligible checks

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > defaults to Pulse retention for non-active subscriptions
[Cleanup] Complete: deleted 1 checks total

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > continues past non-deletable page to delete eligible pulse checks
[Cleanup] Starting tier-aware check cleanup...

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > continues past non-deletable page to delete eligible pulse checks
[Cleanup] Batch 1: deleting 0/1000 eligible checks

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > continues past non-deletable page to delete eligible pulse checks
[Cleanup] Batch 2: deleting 1/2 eligible checks

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > continues past non-deletable page to delete eligible pulse checks
[Cleanup] Complete: deleted 1 checks total

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > deletes orphaned checks when monitor was removed
[Cleanup] Starting tier-aware check cleanup...

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > deletes orphaned checks when monitor was removed
[Cleanup] Batch 1: deleting 1/1 eligible checks

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > deletes orphaned checks when monitor was removed
[Cleanup] Complete: deleted 1 checks total

 ✓ convex/__tests__/monitoring.test.ts (21 tests) 164ms

 Test Files  1 passed (1)
      Tests  21 passed (21)
   Start at  20:46:40
   Duration  870ms (transform 106ms, setup 52ms, import 49ms, tests 164ms, environment 494ms).

File: .groom/retro.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# Delivery Retro Log
+
+## 2026-03-03 - Issue #108
+
+- predicted: effort/s
+- actual: effort/m
+- scope: added tier-aware retention in cleanup action + shared `historyDays` constant + retention tests
+- blocker: pre-push failed initially due pulse interval mismatch in new test; fixed by using tier-valid interval
```
[the remark this answers is not in the document]
Comment (phrazzld) [reply .groom/retro.md]: Classification: style
Severity: low
Decision: fix now. Typo correction is safe and straightforward.
Change: Updated retro note text from "due pulse" to "due to pulse" in  (commit 24af278).
Verification: ; file diff review.

File: convex/monitoring.ts [hunk 4]
```diff
@@ -403,20 +420,72 @@ export const cleanupOldChecks = internalAction({
...
+      const monitors = await ctx.runQuery(
+        internal.monitoring.getMonitorsByIds,
+        {
+          monitorIds,
+        },
+      );
+      const monitorsById = new Map(
+        monitors.map((monitor) => [monitor._id, monitor]),
+      );
+
```
[the remark this answers is not in the document]
Comment (phrazzld) [reply convex/monitoring.ts]: Classification: risk
Severity: medium
Decision: fix now. Single-action N+1 cross-function query pattern reduced.
Change: Added  and replaced per-user  loop in cleanup with one batched query call (, , commit 24af278).
Verification: ; 
 RUN  v4.0.18 /Users/phaedrus/Development/heartbeat

stdout | convex/__tests__/monitoring.test.ts > updateMonitorStatus > third failure sets counter to 3
[Notification] No settings for user user_test, skipping

stdout | convex/__tests__/monitoring.test.ts > updateMonitorStatus > success resets failures even from high count
[Notification] No settings for user user_test, skipping

stdout | convex/__tests__/monitoring.test.ts > updateMonitorStatus > success resets failures even from high count
[Notification] No settings for user user_test, skipping

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > keeps 45-day checks for active Vital subscriptions
[Cleanup] Starting tier-aware check cleanup...

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > keeps 45-day checks for active Vital subscriptions
[Cleanup] Batch 1: deleting 0/1 eligible checks
[Cleanup] Complete: deleted 0 checks total

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > deletes 45-day checks for Pulse subscriptions
[Cleanup] Starting tier-aware check cleanup...

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > deletes 45-day checks for Pulse subscriptions
[Cleanup] Batch 1: deleting 1/1 eligible checks

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > deletes 45-day checks for Pulse subscriptions
[Cleanup] Complete: deleted 1 checks total

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > defaults to Pulse retention for non-active subscriptions
[Cleanup] Starting tier-aware check cleanup...

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > defaults to Pulse retention for non-active subscriptions
[Cleanup] Batch 1: deleting 1/1 eligible checks

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > defaults to Pulse retention for non-active subscriptions
[Cleanup] Complete: deleted 1 checks total

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > continues past non-deletable page to delete eligible pulse checks
[Cleanup] Starting tier-aware check cleanup...

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > continues past non-deletable page to delete eligible pulse checks
[Cleanup] Batch 1: deleting 0/1000 eligible checks

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > continues past non-deletable page to delete eligible pulse checks
[Cleanup] Batch 2: deleting 1/2 eligible checks

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > continues past non-deletable page to delete eligible pulse checks
[Cleanup] Complete: deleted 1 checks total

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > deletes orphaned checks when monitor was removed
[Cleanup] Starting tier-aware check cleanup...

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > deletes orphaned checks when monitor was removed
[Cleanup] Batch 1: deleting 1/1 eligible checks

stdout | convex/__tests__/monitoring.test.ts > cleanupOldChecks > deletes orphaned checks when monitor was removed
[Cleanup] Complete: deleted 1 checks total

 ✓ convex/__tests__/monitoring.test.ts (21 tests) 151ms

 Test Files  1 passed (1)
      Tests  21 passed (21)
   Start at  20:47:00
   Duration  515ms (transform 94ms, setup 47ms, import 43ms, tests 151ms, environment 205ms).

File: convex/monitoring.ts [hunk 5]
```diff
@@ -383,40 +402,113 @@ export const runHeartbeat = internalAction({
 export const cleanupOldChecks = internalAction({
   args: {},
```
[the remark this answers is not in the document]
Comment (phrazzld) [reply convex/monitoring.ts]: Classification: bug
Severity: critical
Decision: fixed
Change: Addressed in commit e64b2cf. `cleanupOldChecks` now accepts optional `cursor` args and resumes from `args.cursor ?? null` (`convex/monitoring.ts:403`, `convex/monitoring.ts:414`). Rescheduling now forwards progress via `ctx.scheduler.runAfter(..., { cursor })` (`convex/monitoring.ts:517`). Cron registration now passes explicit args object for the updated signature (`convex/crons.ts:10`). Added regression test `resumes cleanup from provided cursor` (`convex/__tests__/monitoring.test.ts:509`).
Verification: `bunx vitest run convex/__tests__/monitoring.test.ts`, `bun lint`, `bun type-check`
